activerecord/lib/active_record/relation.rb

```diff
@@ -315,9 +315,16 @@ def scoping
       klass.current_scope = previous
     end
 
-    def _exec_scope(*args, &block) # :nodoc:
+    def _exec_scope(name, *args, &block) # :nodoc:
       @delegate_to_klass = true
-      instance_exec(*args, &block) || self
+      instance_exec(*args, &block).tap do |result|
+        valid_scope_result = result.is_a?(Relation) || !result
+        unless valid_scope_result
+          logger&.warn "#{klass}.#{name} returned a value of type #{result.class}. " \
```

Comment:
Can we deprecate this and raise in 6.2 version? This warning will be completely ignored.

Reply:
I'd really like that as it will save other souls from it. I'll push the changes in a bit and we can check whether the deprecation message is explicit enough.